internal/indexer/oci.go

```diff
@@ -0,0 +1,204 @@
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/bitnami-labs/charts-syncer/internal/indexer/api"
+	"github.com/bitnami-labs/pbjson"
+	containerderrs "github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes"
+	"github.com/pkg/errors"
+	"oras.land/oras-go/pkg/content"
+	"oras.land/oras-go/pkg/oras"
+)
+
+// ociIndexerOpts are the options to configure the ociIndexer
+type ociIndexerOpts struct {
+	reference string
+	url      string
+	username string
+	password string
+	insecure bool
+}
+
+// OciIndexerOpt allows setting configuration options
+type OciIndexerOpt func(opts *ociIndexerOpts)
+
+// WithIndexRef configures the charts index OCI reference instead of letting the library
+// using the default host/index:latest one.
+//
+// 	opt := WithIndexRef("my.oci.domain/index:prod")
+//
+func WithIndexRef(r string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.reference = r
+	}
+}
+
+// WithBasicAuth configures basic authentication for the OCI host
+//
+// 	opt := WithBasicAuth("user", "pass")
+//
+func WithBasicAuth(user, pass string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.username = user
+		opts.password = pass
+	}
+}
+
+// WithInsecure configures insecure connection
+//
+// 	opt := WithInsecure()
+//
+func WithInsecure() OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.insecure = true
+	}
+}
+
+// WithHost configures the OCI host
+//
+// 	opt := WithHost("my.oci.domain")
+//
+func WithHost(h string) OciIndexerOpt {
+	return func(opts *ociIndexerOpts) {
+		opts.url = h
+	}
+}
+
+// ociIndexer is an OCI-based Indexer
+type ociIndexer struct {
+	reference string
+	resolver  remotes.Resolver
+}
+
+// DefaultIndexName is the name for the OCI artifact with the index
+const DefaultIndexName = "index"
+
+// DefaultIndexTag is the tag for the OCI artifact with the index
+const DefaultIndexTag = "latest"
+
+// NewOciIndexer returns a new OCI-based indexer
+func NewOciIndexer(opts ...OciIndexerOpt) (Indexer, error) {
+	opt := &ociIndexerOpts{}
+	for _, o := range opts {
+		o(opt)
+	}
+
+	u, err := url.Parse(opt.url)
+	if err != nil {
+		return nil, errors.Wrapf(ErrInvalidArgument, "invalid OCI host URL: %+v", err)
+	}
+	resolver := newDockerResolver(u, opt.username, opt.password, opt.insecure)
+
+	ind := &ociIndexer{
+		reference: opt.reference,
+		resolver:  resolver,
+	}
+	if ind.reference == "" {
+		uri := strings.Trim(strings.Join([]string{u.Host, u.Path}, "/"), "/")
+		ind.reference = fmt.Sprintf("%s/%s:%s", uri, DefaultIndexName, DefaultIndexTag)
+		klog.Infof("Remote index was not provided, using default location: %q", ind.reference)
+	} else {
+		klog.Infof("Remote index was provided: %q", ind.reference)
+	}
+
+	return ind, nil
+}
+
+// VacAssetIndexLayerMediaType is a media type used in VAC to store a JSON containing the index of
+// charts and containers in a repository
+const VacAssetIndexLayerMediaType = "application/vnd.vmware.tac.asset-index.layer.v1.json"
```

Comment:
do we really need to depend on this media type? I'd not like to have a VAC reference here if it's not strictly needed.

Reply:
So far, it is the only way of controlling the layer that defines the index content and filename. Ideally, I would not do this either. See my other comment.